### ISSUE TITLE
[NCC-E005955-6AN] `zebra-network`: Power-of-Two-Choices Load Balancing May Deprioritize Honest Peers

### ISSUE THREAD

**Issue by mpguerra** [ZcashFoundation/zebra#6343]
## Details

The “power of two choices” load-balancing algorithm routes outbound requests by first randomly sampling a set of known peers and then sending the request to the lowest-latency peer in the set. While this clearly behaves well under normal conditions and has attractive worst-case bounds, we note that it also introduces risks against attackers with significant influence over the network. An ISP-level attacker, for instance, could deploy peers with exceptionally low latency, and could even artificially inflate other peers’ latencies, thereby increasing their level of influence over the network. Such an attack could be difficult to detect.

For instance, in the case of an attempted eclipse attack, suppose a peer’s set of outbound connections is 80% compromised, with only 20% of their remote peers being honest. Ordinarily, such a peer would still be connected to the (honest part of the) network, and would not be considered compromised. However, suppose further that using p2c routing, this peer chooses two remote peers as candidates for an outbound request. The likelihood of both these peers being honest is only 4%, and in all other cases we assume the attacker’s low-latency remote peer will be prioritized over a randomly sampled honest peer.

Thus, even though the peer retains honest connections, it may deprioritize these honest peers to the point where the net effect is comparable to full compromise.

There are several ways this might be mitigated; as one idea, perhaps latencies below a certain dynamically-determined threshold (e.g. median latency) could be treated as equivalent, and the second choice could fall back to random selection as a tiebreaker. This would retain the desired behavior of deprioritizing abnormally high-latency peers while also
avoiding prioritization of abnormally low-latency peers.

**Comment by teor2345** [comment]
> An ISP-level attacker, for instance, could deploy peers with exceptionally low latency, and could even artificially inflate other peers’ latencies, thereby increasing their level of influence over the network. Such an attack could be difficult to detect.

An attacker that can place peers in the same data center can also perform this attack, most cloud services allow data center selection.

**Comment by teor2345** [comment]
> There are several ways this might be mitigated; as one idea, perhaps latencies below a certain dynamically-determined threshold (e.g. median latency) could be treated as equivalent, and the second choice could fall back to random selection as a tiebreaker. This would retain the desired behavior of deprioritizing abnormally high-latency peers while also avoiding prioritization of abnormally low-latency peers.

This is a good suggestion, because it would only slow down 1/4 of requests (`50%` of peers are below the median, so `50% * 50% = 25%` of randomly selected pairs of peers are both below the median).

However, it would not reduce the impact of an 80% attack with low-latency peers, because the median would be a low-latency peer. (It would only reduce the impact of attacks below 50%, and only by the proportion of honest peers below the median.)

**Comment by teor2345** [comment]
Using the arithmetic mean would defend against all abnormally low-latency peers, but at the cost of slowing down a larger proportion of requests in the non-attack case.

(We'd also need to think about how to calculate this average efficiently, if we query all the peers for their load on every request, then we'll be making many calls to `Instant::now()`, which is slow on some machines. So we might want to cache the latest load for a peer each time we query it, and use that for the average.)

**Comment by teor2345** [comment]
I've thought about this a bit more, and I don't think there is an algorithm that's clearly better that we could switch to now.

But we could look for ways to harden the peer selection parts of `zebra-network`, it's possible that some minor changes could significantly improve security.

I also don't think this is a significant attack on Zebra right now, because we connect to 75 peers under load, so it's more difficult to set up a large number of peers to do a statistical attack.

**Comment by mpguerra** [comment]
What are the next steps here? Will we just close as "won't-fix" for now?

**Comment by teor2345** [comment]
> What are the next steps here? Will we just close as "won't-fix" for now?

I think so.

We could lower the number of inbound connections we allow, but that risks another kind of denial of service. (Where peers that can only make outbound connections don't have any nodes to connect to.)

I can't think of any other easy hardening step, we've implemented some already: peer limits, inventory targeting, randomised peer selection.

So to exploit this bug, a malicious group of peers has to:
1. Become most of Zebra's 50-200 peers in the peer set
2. Provide fake peer addresses, blocks, or transactions

Providing fake blocks is the biggest risk, to do that, peers have to:
1. Mine valid fake blocks
2. Get those block hashes in Zebra's inventory
3. Get selected for those blocks
4. Provide those blocks
5. Consistently repeat steps 1-4 around 100 times, at a greater rate than the majority of the peers on the network

We've also hardened peer address selection, by rejecting addresses with fake times. So eventually all peers get tried.
